server/src/main/java/org/elasticsearch/cluster/service/MasterService.java

```diff
@@ -711,6 +711,14 @@ public void onCommit(TimeValue commitTime) {
                 assert false : "ackTimeout must always be present: " + contextPreservingAckListener;
                 ackTimeout = TimeValue.ZERO;
             }
+
+            if (ackTimeout.millis() < 0) {
+                if (countDown.countDown()) {
+                    finish();
+                }
+                return;
```

Comment:
The sequence is right.

You could be right about `onCommit` never actually finishing the acking today. FWIW this code was added in #31303 (6.4.0) and we've rewritten much of the surrounding code since then. That said, it's a fairly delicate argument to prove this, whereas the "obviously correct" code as written today is robust and not meaningfully less efficient.

In particular, it's hard to see that `onCommit` is called before any `ApplyCommit` is sent. The relevant code is `org.elasticsearch.cluster.coordination.Publication.PublicationTarget#handlePublishResponse`:

```
        void handlePublishResponse(PublishResponse publishResponse) {
            assert isWaitingForQuorum() : this;
            logger.trace("handlePublishResponse: handling [{}] from [{}])", publishResponse, discoveryNode);
            if (applyCommitRequest.isPresent()) {
                sendApplyCommit();
            } else {
                try {
                    Publication.this.handlePublishResponse(discoveryNode, publishResponse).ifPresent(applyCommit -> {
                        assert applyCommitRequest.isPresent() == false;
                        applyCommitRequest = Optional.of(applyCommit);
                        ackListener.onCommit(TimeValue.timeValueMillis(currentTimeSupplier.getAsLong() - startTime));
                        publicationTargets.stream()
                            .filter(PublicationTarget::isWaitingForQuorum)
                            .forEach(PublicationTarget::sendApplyCommit);
                    });
                } catch (Exception e) {
                    setFailed(e);
                    onPossibleCommitFailure();
                }
            }
        }
```

As written, you could have the committing thread setting `applyCommitRequest`, then pausing before calling `ackListener.onCommit`, while another thread concurrently processes a later `PublishResponse`, discovers that `applyCommitRequest.isPresent()` and sends the `ApplyCommit`. But then if you look at how this code is called eventually you discover that this all happens underneath `Coordinator#mutex` so these things cannot happen concurrently. But relying on a mutex in `Coordinator` to protect against concurrency in `Publication` as part of the correctness argument for `MasterService` is too deeply opaque for my liking.

Reply:
Thanks a lot for the explanation. TIL 🙇